### PR TITLE
planner/core, util/plancodec: fix wrongly encoded plan which might cause panic

### LINKED ISSUE #36609: panic when select a complex plan from statements_summary
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
use test;
create table t1(a int, b int, c int, d int, index ia(a), index ib(b), index ic(c), index id(d));
create table t2(a int, b int, c int, d int, index ia(a), index ib(b), index ic(c), index id(d));
create table t3(a int, b int, c int, d int, index ia(a), index ib(b), index ic(c), index id(d));
create table t4(a int, b int, c int, d int, index ia(a), index ib(b), index ic(c), index id(d));
create table t5(a int, b int, c int, d int, index ia(a), index ib(b), index ic(c), index id(d));
select * from t3 straight_join t4 on t3.a = t4.b straight_join t2 on t3.d = t2.c straight_join t1 on t1.a = t2.b straight_join t5 on t4.c = t5.d where t2.b < 100 and t4.a = 10;
select * from information_schema.statements_summary;
```

### 2. What did you expect to see? (Required)

The query success.

### 3. What did you see instead (Required)
```
> select * from information_schema.statements_summary;
ERROR 1105 (HY000): runtime error: index out of range [6] with length 4
```
### 4. What is your TiDB version? (Required)

v6.2, master

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #36609

Problem Summary:

When encoding the plan and meeting an operator whose `NeedReverseDriverSide` is true, we get the wrong build side subtree.

### What is changed and how it works?

- Correct the wrong subtree logic.
- Add `recover()` to `DecodePlan` to avoid panic.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

File: util/plancodec/codec.go [hunk 1]
```diff
@@ -65,7 +65,12 @@ var decoderPool = sync.Pool{
 }
 
 // DecodePlan use to decode the string to plan tree.
-func DecodePlan(planString string) (string, error) {
+func DecodePlan(planString string) (res string, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.New("DecodePlan panicked")
```
Comment (hawkingrei) [inline util/plancodec/codec.go:71]: We need to print the panic stack to show how this panic happens.

Comment (AilinKid) [reply util/plancodec/codec.go:71]: debug.PrintStack()

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: util/plancodec/codec.go (hunk 1 above)
Comment (hawkingrei) [reply util/plancodec/codec.go:71]: zap.Any("recover", r)

Comment (time-and-fate) [reply util/plancodec/codec.go:74]: Updated. PTAL~

**Review by AilinKid — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**